sdk/tools/asmpp/asmpp.cpp

```diff
@@ -0,0 +1,1152 @@
+
+            if ((tokens[index].type() == TOKEN_TYPE::WhiteSpace) &&
+                (tokens[index + 1].type() == TOKEN_TYPE::KW_FRAME))
+            {
+#ifdef TARGET_amd64
+                printf("\n.seh_proc %s\n", tok.str().c_str());
+#else
+                printf("\n.cfi_startproc\n");
+#endif
+                index += 2;
+            }
+            break;
+        }
+
+        case TOKEN_TYPE::KW_ENDP:
+        {
+            printf(".seh_endproc\n.endfunc");
+            index += 3;
+            break;
+        }
+
+        case TOKEN_TYPE::KW_RECORD:
+            index = translate_record(tokens, index, macro_params);
+            break;
+
+        default:
+            // We don't know what it is, assume it's a macro and trat it like an instruction
+            index = translate_instruction(tokens, index, macro_params);
+            break;
+    }
+
+    return index;
+}
+
+size_t
+translate_construct(TokenList& tokens, size_t index, const vector<string> &macro_params)
+{
+    Token tok = tokens[index];
+
+    switch (tok.type())
+    {
+        case TOKEN_TYPE::WhiteSpace:
+        case TOKEN_TYPE::NewLine:
+        case TOKEN_TYPE::Comment:
+            return translate_token(tokens, index, macro_params);
+
+        case TOKEN_TYPE::Identifier:
+            return translate_identifier_construct(tokens, index, macro_params);
+
+        case TOKEN_TYPE::KW_ALIGN:
+            index = translate_construct_one_param(".align", tokens, index, macro_params);
+            break;
+
+        case TOKEN_TYPE::KW_allocstack:
+            index = translate_construct_one_param(".seh_stackalloc", tokens, index, macro_params);
+            break;
+
+        case TOKEN_TYPE::KW_code:
+#ifdef TARGET_amd64
+            printf(".code64");
+#else
+            printf(".code");
+#endif
+            printf(" .intel_syntax noprefix");
+            index++;
+            break;
+
+        case TOKEN_TYPE::KW_const:
+            printf(".section .rdata");
+            index++;
+            break;
+
+        case TOKEN_TYPE::KW_DB:
+        case TOKEN_TYPE::KW_DW:
+        case TOKEN_TYPE::KW_DD:
+        case TOKEN_TYPE::KW_DQ:
+            return translate_data_def(tokens, index, macro_params);
+
+        case TOKEN_TYPE::KW_END:
+            printf("// END\n");
+            return tokens.size();
+
+        case TOKEN_TYPE::KW_endprolog:
+            printf(".seh_endprologue");
+            index++;
+            break;
+
+        case TOKEN_TYPE::KW_EXTERN:
+        {
+            Token tok1 = get_ws_or_nl(tokens[index + 1]);
+            Token tok2 = get_expected_token(tokens[index + 2], TOKEN_TYPE::Identifier);
+            add_mem_id(tok2);
+            printf("//");
+            return complete_line(tokens, index, macro_params);
+        }
+
+        case TOKEN_TYPE::KW_if:
+        case TOKEN_TYPE::KW_ifdef:
+        case TOKEN_TYPE::KW_ifndef:
+        case TOKEN_TYPE::KW_else:
+        case TOKEN_TYPE::KW_endif:
+            // FIXME: not really good
```

Comment:
What would be one better solution?

Reply:
It's a preliminary solution that works with the code I have, but there are some differences in how MASM"if" and GAS ".if" work with the parameters. I'll improve the comment to make that more clear.